### PR TITLE
Add design docs

### LINKED ISSUE #1293: Standard API for major event handlers in 2.0
While we're writing components for 2.0, we should standardize the API got event handlers.

Currently with props all event handlers are native event handlers.
E.g: `onChange` of a checkbox is the native onChange of `input type=checkbox`
This leads to some problems, mdc-web introduce some additional properties on top of native controls.
E.g. checkbox has a `indeterminate` state, select has `selectedIndex` etc.

Getting these values from native event handlers which just pass the `Event`, is tricky and most times requires creating a ref.

In order to solve this we can send added details to major event handlers(not necessarily all) like onChange for checkbox, select, onClick for button etc.

Current TS definition:
```
onClick: (e:Event) => void;
```

Updated version:

```
onChange:(e:Event, object: {
  MDComponent: MaterialComponent;
  checked: boolean;
  indeterminate: boolean;
}) => void
```

The additional object makes sure that it is not counter intuitive to users if 1.0

### PR DESCRIPTION
**Pull request by cromefire**
Fixes #1293

### CODE REVIEW

**Comment by cromefire** [comment]
@prateekbh Do you miss anything?

**Comment by artipost** [comment]
Failed pictures (27): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/bf86ba20f222d816d6806d7b33d28061/raw/7744e66234b574b5adc2239843919f3664453df3/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [502607770](https://travis-ci.org/prateekbh/preact-material-components/builds/502607770), for commit [69cb4b9](https://github.com/prateekbh/preact-material-components/pull/1296/commits/69cb4b9266f56668e24d0bde67ed9665b3a986c8).

**Comment by artipost** [comment]
Failed pictures (27): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/cdc4d5b556a90ac39be99e0a68ea3e7c/raw/f40ca6f1ca5612137e1cd636edb9c180a13ac4b3/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [502610544](https://travis-ci.org/prateekbh/preact-material-components/builds/502610544), for commit [3378629](https://github.com/prateekbh/preact-material-components/pull/1296/commits/337862952741a2901ffb8ee2b8a47cc5dfcd2ab4).

**Comment by artipost** [comment]
Failed pictures (27): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/4a801fd7602677e1e1a329c41ccd21ba/raw/6a156c340b792517e01e0e7dff2f4f86749e3da1/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [502612621](https://travis-ci.org/prateekbh/preact-material-components/builds/502612621), for commit [f314011](https://github.com/prateekbh/preact-material-components/pull/1296/commits/f3140113c334f2d3ce4b0a261f0c2006944bbfa0).

**Comment by artipost** [comment]
Failed pictures (50): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/1e652d7f51c1fe45bca59cdf4f36a239/raw/9c5dc0bd47c060e327e4bfc939ab7e1f1984ba37/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [502617987](https://travis-ci.org/prateekbh/preact-material-components/builds/502617987), for commit [1fe0a04](https://github.com/prateekbh/preact-material-components/pull/1296/commits/1fe0a045b1e61dc47ed233f421e92e14616e0a34).

File: packages/docs/src/components/code-block/index.js [hunk 1]
```diff
@@ -3,12 +3,14 @@ import {h} from 'preact';
 import cx from 'classnames';
 import hljs from 'highlight.js/lib/highlight';
 import javascript from 'highlight.js/lib/languages/javascript';
+import typescript from 'highlight.js/lib/languages/typescript';
 import htmlbars from 'highlight.js/lib/languages/htmlbars';
 import json from 'highlight.js/lib/languages/json';
 import xml from 'highlight.js/lib/languages/xml';
+import python from 'highlight.js/lib/languages/python';
```
Comment (prateekbh) [inline packages/docs/src/components/code-block/index.js]: This is a very one off use case, do we know how bug is this? If big can we load this lazily only for that component?

**Comment by prateekbh** [comment]
Can u update this PR to remove the screenshots

**Comment by cromefire** [comment]
Sure I can tomorrow

File: packages/docs/src/components/code-block/index.js (hunk 1 above)
Comment (cromefire) [reply packages/docs/src/components/code-block/index.js]: I have to admit it's may be a bit wast full I maybe wanted to write a own "language" I only need the hashtag comment and that should ba a one-liner

File: packages/docs/src/routes/design/ev-interface.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+interface IEventHandler {
+    component: MaterialComponent;
```
Comment (prateekbh) [inline packages/docs/src/routes/design/ev-interface.txt]: Component or MDComponent

Comment (cromefire) [reply packages/docs/src/routes/design/ev-interface.txt]: Hmm I guess I meant `MDComponent`

Comment (cromefire) [reply packages/docs/src/routes/design/ev-interface.txt]: No I actually meant `MaterialComponent` the super class of all our components. Improved the docs for that

File: packages/docs/src/components/code-block/index.js (hunk 1 above)
Comment (cromefire) [reply packages/docs/src/components/code-block/index.js]: Added that simple language

**Pull request merged**